### PR TITLE
Full UFS support

### PR DESCRIPTION
**Pull request by BotchedRPR**
With this PR, we achieve full UFS RW on Exynos990 platforms

Checklist:

- [x] UFS probes
- [x] We get a a link
- [x] Can read/write blocks fine
- [x] Can read/write partitions with PIT

Thanks!

### CODE REVIEW

File: engine.mk [hunk 1]
```diff
@@ -177,7 +177,7 @@ endif
 
 #build mode defines
 ifeq ($(USER), user)
-GLOBAL_DEFINES += RAMDUMP_MODE_OFF
+GLOBAL_DEFINES += RAMDUMP_MODE_OFF LK3RD_USER
```
Comment (ivoszbg) [inline engine.mk]: How about we move that out to the platform rules.mk? As-in LK3RD_USER=1

Comment (BotchedRPR) [reply engine.mk]: The debug logger commit was a very-much temporary measure -  but sure, if I decide to merge it.
Acked

Comment (BotchedRPR) [reply engine.mk]: Commit dropped.

**Comment by halal-beef** [comment]
1 favour, please do tick off or remove the todo in the readme, review coming very very soon

**Comment by halal-beef** [comment]
https://github.com/exynos990-mainline/lk3rd/commit/a39ed9c3337236b7933a0cb8a88c8bf6b7aebc17#diff-5467df3a559afa4d9dd401e99dac28abec7c8fa5c48bb4ec263fac558df7ae1bR473

was this meant to be here?

https://github.com/exynos990-mainline/lk3rd/commit/a39ed9c3337236b7933a0cb8a88c8bf6b7aebc17#diff-232990cb7a430da1bb24b1ab68f652ceb4cb61e50f76437df7bbaa003338c3cbR188

also this?

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L26

wasnt this in a define

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L70

act upon the TODO or remove

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L135

is this really appropiate to be in here?

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L185

reword

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L235

reword

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L311C1-L317C40

there really should be that userdata fmp part in here imo

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L388

reword

https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L433C1-L439C2

should be completed imo, but for functionality right now, this lib at this state is fine

**Comment by halal-beef** [comment]
other than those tho, really good job

**Review by halal-beef — CHANGES_REQUESTED** [review]
Everything requested above.

**Comment by BotchedRPR** [comment]
> https://github.com/exynos990-mainline/lk3rd/blob/d28cf6aaeb795174897feef8376a4c23d8d93a32/lib/gpt/pit.c#L26
> 
> wasnt this in a define

We could use PIT_DISK_SIZE_LIMIT, but that isn't the initial block count for PIT. 
Other than that, I do not see any defines for the initial block count.

**Comment by BotchedRPR** [comment]
All changes should be sorted. Thank you for the review.

I could tick off the README checklist after the merge. My branch is older, so I can't directly do it.

**Review by halal-beef — APPROVED** [review]
All good over here, we should probably get sparse up in the future tho

**Pull request merged**